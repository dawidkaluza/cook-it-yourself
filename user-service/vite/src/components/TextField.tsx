import {ChangeEventHandler, useId} from "react";

export const TextField = (props: {
  name: string;
  label: string;
  type?: "text" | "password";
  value?: string;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  placeholder?: string;
  style?: string;
}) => {
  const id = useId();
  return (
    <div className={props.style}>
      <label htmlFor={props.name + "_" + id} className="block mb-1 font-medium">
        {props.label}
      </label>
      <input
        id={props.name + "_" + id}
        name={props.name}
        type={props.type}
        value={props.value}
        onChange={props.onChange}
        placeholder={props.placeholder}
        className="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
      />
    </div>
  );
};