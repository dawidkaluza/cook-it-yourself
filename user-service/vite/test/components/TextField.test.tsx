import {render, screen} from "@testing-library/react";
import {TextField} from "../../src/components/TextField.tsx";
import {useState} from "react";
import {expect} from "vitest";
import {userEvent} from "@testing-library/user-event";

vi.mock("react", async function <T>(importOriginal: () => Promise<T>) {
  const original = await importOriginal();
  return {
    ...original,
    useId: () => "1234",
  };
});

const ControlledNameField = () => {
  const [name, setName] = useState("");

  return (
    <TextField
      name={"name"}
      label={"Name"}
      value={name}
      onChange={(event) => setName(event.target.value)}
    />
  );
};

describe("TextField component", () => {
  test("render with required props", () => {
    // Given
    // When
    const component = render(<TextField name="name" label="Name" onChange={() => {}} />);

    // Then
    const inputEl = screen.getByLabelText("Name");
    expect(inputEl).toBeDefined();
    expect(inputEl.getAttribute("name")).toBe("name");
    expect(inputEl.getAttribute("type")).toBeNull();
    expect(inputEl.getAttribute("value")).toBe("");
    expect(component).toMatchSnapshot();
  });

  test.each([
    ["text"],
    ["password"],
  ])("render with type %s and all other props", (type) => {
    // Given
    // When
    const component = render(
      <TextField
        name="name" label="Name"
        type={type as "password" | "text"} value="Dawid"
        error="Invalid value"
        onChange={() => {}}
        placeholder="Dawid" style="w-full"
      />
    );

    // Then
    const inputEl = screen.getByLabelText("Name");
    expect(inputEl).toBeDefined();
    expect(inputEl.getAttribute("name")).toBe("name");
    expect(inputEl.getAttribute("type")).toBe(type);
    expect(inputEl.getAttribute("value")).toBe("Dawid");
    expect(inputEl.getAttribute("placeholder")).toBe("Dawid");

    const errorEl = screen.getByText("Invalid value");
    expect(errorEl).toBeDefined();

    const containerEl = inputEl.closest("div") as HTMLDivElement;
    expect(containerEl).not.toBeNull();
    expect(containerEl.getAttribute("class")).toContain("w-full");

    expect(component.container).toMatchSnapshot();
  });

  test("render controlled and change its value", async () => {
    // Given
    const user = userEvent.setup();
    const component = render(<ControlledNameField />);
    const inputEl = screen.getByLabelText("Name");

    // When
    await user.click(inputEl);
    await user.keyboard("foo");

    // Then
    expect(inputEl.getAttribute("value")).toBe("foo");
    expect(component.container).toMatchSnapshot();
  });
})