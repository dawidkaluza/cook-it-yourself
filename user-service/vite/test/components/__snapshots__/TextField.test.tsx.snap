// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField component > render controlled and change its value 1`] = `
<div>
  <div>
    <label
      class="block mb-1 font-medium"
      for="name_1234"
    >
      Name
    </label>
    <input
      class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
      id="name_1234"
      name="name"
      value="foo"
    />
  </div>
</div>
`;

exports[`TextField component > render with required props 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <label
          class="block mb-1 font-medium"
          for="name_1234"
        >
          Name
        </label>
        <input
          class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
          id="name_1234"
          name="name"
          value=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <label
        class="block mb-1 font-medium"
        for="name_1234"
      >
        Name
      </label>
      <input
        class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400"
        id="name_1234"
        name="name"
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TextField component > render with type password and all other props 1`] = `
<div>
  <div
    class="w-full"
  >
    <label
      class="block mb-1 font-medium"
      for="name_1234"
    >
      Name
    </label>
    <input
      class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400 border-red-600"
      id="name_1234"
      name="name"
      placeholder="Dawid"
      type="password"
      value="Dawid"
    />
    <span
      class="block text-xs text-red-600 "
    >
      Invalid value
    </span>
  </div>
</div>
`;

exports[`TextField component > render with type text and all other props 1`] = `
<div>
  <div
    class="w-full"
  >
    <label
      class="block mb-1 font-medium"
      for="name_1234"
    >
      Name
    </label>
    <input
      class="w-full py-2 px-3 rounded border shadow text-gray-700 outline-none focus:border-blue-400 border-red-600"
      id="name_1234"
      name="name"
      placeholder="Dawid"
      type="text"
      value="Dawid"
    />
    <span
      class="block text-xs text-red-600 "
    >
      Invalid value
    </span>
  </div>
</div>
`;
